api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+		ticker:        time.NewTicker(config.RetryWaitFrequency),
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterSendFilServiceServer(s.server, s)
+		if err := s.server.Serve(config.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	if err := s.waitAllQualifying(ctx, true); err != nil {
+		return nil, fmt.Errorf("calling waitAllQualifying: %v", err)
+	}
+
+	s.bindTicker(ctx)
+
+	return s, nil
+}
+
+func (s *Service) waitAllQualifying(ctx context.Context, isInitialRun bool) error {
+	filter := bson.M{"$and": bson.A{
+		bson.M{"message_state": bson.M{"$ne": pb.MessageState_MESSAGE_STATE_ACTIVE}},
+		bson.M{"message_state": bson.M{"$ne": pb.MessageState_MESSAGE_STATE_FAILED}},
+	}}
+	if !isInitialRun {
+		filter["waiting"] = false
+	}
+	cursor, err := s.col.Find(ctx, filter)
+	if err != nil {
+		return status.Errorf(codes.Internal, "querying txns: %v", err)
+	}
+	defer cursor.Close(ctx)
+	var txns []txn
+	err = cursor.All(ctx, &txns)
+	if err != nil {
+		return status.Errorf(codes.Internal, "decoding txns query results: %v", err)
+	}
+	log.Infof("found %v txns to initiate waiting on", len(txns))
+	for _, txn := range txns {
+		s.wait(txn)
+	}
+	return nil
+}
+
+func (s *Service) bindTicker(ctx context.Context) {
+	go func() {
+		for {
+			<-s.ticker.C
+			if err := s.waitAllQualifying(ctx, false); err != nil {
+				log.Errorf("waitAllQualifying from ticker: %v", err)
+			}
+		}
+	}()
+}
+
+func (s *Service) SendFil(ctx context.Context, req *pb.SendFilRequest) (*pb.SendFilResponse, error) {
+	f, err := address.NewFromString(req.From)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing from address: %v", err)
+	}
+	t, err := address.NewFromString(req.To)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing to address: %v", err)
+	}
+	nanoAmount := (&big.Int{}).SetInt64(req.AmountNanoFil)
+	factor := (&big.Int{}).SetInt64(int64(math.Pow10(9)))
+	amount := (&big.Int{}).Mul(nanoAmount, factor)
+	msg := &types.Message{
+		From:  f,
+		To:    t,
+		Value: types.BigInt{Int: amount},
+	}
+	client, cls, err := s.clientBuilder(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "creating filecoin client: %v", err)
+	}
+	defer cls()
+
+	sm, err := client.MpoolPushMessage(ctx, msg, nil)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "pushing message: %v", err)
+	}
+
+	now := time.Now()
+
+	tx := txn{
+		ID:            primitive.NewObjectID(),
+		From:          req.From,
+		To:            req.To,
+		AmountNanoFil: req.AmountNanoFil,
+		MessageCids:   []msgCid{{Cid: sm.Message.Cid().String(), CreatedAt: now}},
+		MessageState:  pb.MessageState_MESSAGE_STATE_PENDING,
+		CreatedAt:     now,
+		UpdatedAt:     now,
+	}
+
+	if _, err = s.col.InsertOne(ctx, &tx); err != nil {
+		return nil, status.Errorf(codes.Internal, "inserting txn into collection: %v", err)
+	}
+
+	wait := s.wait(tx)
```

Comment:
We always call `wait` when `SendFil` is called. `wait` returns a channel that may or may not be read from.

Reply:
This is resolved in later comments about side effects etc. This will change in the next PR when I implement waiting and notifying properly.